### ISSUE TITLE
ValueError: as_list() is not defined on an unknown TensorShape.

### ISSUE THREAD

**Issue by dundarahmet** [tensorflow/tensorflow#49238]
Hi. I try to implement AlexNet with Coco dataset. I want to make multi label classification but tf throws the error. Full error is:

`
ValueError: in user code:

    /usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/engine/training.py:805 train_function  *
        return step_function(self, iterator)
    /usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/engine/training.py:795 step_function  **
        outputs = model.distribute_strategy.run(run_step, args=(data,))
    /usr/local/lib/python3.7/dist-packages/tensorflow/python/distribute/distribute_lib.py:1259 run
        return self._extended.call_for_each_replica(fn, args=args, kwargs=kwargs)
    /usr/local/lib/python3.7/dist-packages/tensorflow/python/distribute/distribute_lib.py:2730 call_for_each_replica
        return self._call_for_each_replica(fn, args, kwargs)
    /usr/local/lib/python3.7/dist-packages/tensorflow/python/distribute/distribute_lib.py:3417 _call_for_each_replica
        return fn(*args, **kwargs)
    /usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/engine/training.py:788 run_step  **
        outputs = model.train_step(data)
    /usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/engine/training.py:758 train_step
        self.compiled_metrics.update_state(y, y_pred, sample_weight)
    /usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/engine/compile_utils.py:387 update_state
        self.build(y_pred, y_true)
    /usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/engine/compile_utils.py:318 build
        self._metrics, y_true, y_pred)
    /usr/local/lib/python3.7/dist-packages/tensorflow/python/util/nest.py:1163 map_structure_up_to
        **kwargs)
    /usr/local/lib/python3.7/dist-packages/tensorflow/python/util/nest.py:1258 map_structure_with_tuple_paths_up_to
        func(*args, **kwargs) for args in zip(flat_path_gen, *flat_value_gen)
    /usr/local/lib/python3.7/dist-packages/tensorflow/python/util/nest.py:1258 <listcomp>
        func(*args, **kwargs) for args in zip(flat_path_gen, *flat_value_gen)
    /usr/local/lib/python3.7/dist-packages/tensorflow/python/util/nest.py:1161 <lambda>
        lambda _, *values: func(*values),  # Discards the path arg.
    /usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/engine/compile_utils.py:418 _get_metric_objects
        return [self._get_metric_object(m, y_t, y_p) for m in metrics]
    /usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/engine/compile_utils.py:418 <listcomp>
        return [self._get_metric_object(m, y_t, y_p) for m in metrics]
    /usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/engine/compile_utils.py:439 _get_metric_object
        y_t_rank = len(y_t.shape.as_list())
    /usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/tensor_shape.py:1190 as_list
        raise ValueError("as_list() is not defined on an unknown TensorShape.")

    ValueError: as_list() is not defined on an unknown TensorShape.
`

Here the Colab link:
[https://colab.research.google.com/drive/12nO4IDQJg02aiduqsMimyMJ6Z8KNlCFw?usp=sharing](url)

I don't understand where my mistake is.
**System information**
-Google Colab

You can collect some of this information using our environment capture
You can also obtain the TensorFlow version with:
TF 2.0: v2.4.1-0-g85c8b2a817f 2.4.1

**Describe the current behavior**

**Describe the expected behavior**

**[Contributing](https://www.tensorflow.org/community/contribute)** - Do you
want to contribute a PR? (yes/no): - Briefly describe your candidate solution
(if contributing):

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by lonerzzz** [comment]
This was marked as a bug. Is this a new bug or is it tied to the same bugs around make_ndArray() and numpy() methods being missing?

**Comment by bhack** [comment]
Check https://github.com/tensorflow/tensorflow/issues/28257#issuecomment-489717572

**Comment by dundarahmet** [comment]
> Check [#28257 (comment)](https://github.com/tensorflow/tensorflow/issues/28257#issuecomment-489717572)

I've read it but in my situation tf.py_function returns a tuple which includes an image and a label. I cannot give the shape of anything. Here the code:

`ds_train = tf.data.Dataset.from_tensor_slices(train_annotation.index.values).map(lambda x: tf.py_function(train_images, [x], [tf.float32, tf.uint8]))`

actually, tf.py_function's input parameters is just a file name, not an input that is fed into the network.



> This was marked as a bug. Is this a new bug or is it tied to the same bugs around make_ndArray() and numpy() methods being missing?

I don't know. I just want to get image and label from a csv file which inclide the file name and a multi label.

**Comment by dundarahmet** [comment]
update the problem:

I tried to give the shapes to the image and label but this didn't work.  Here is the test codes and the error I got:

 def train_images(fn):
> file_name = bytes.decode(fn.numpy())
  image = tf.io.read_file("train2014/" + file_name)
  image = tf.image.decode_image(image, channels=3, dtype=tf.float32)
  image = tf.image.resize(image, (227, 227))
  label = tf.convert_to_tensor(train_annotation.loc[file_name].values, dtype=tf.uint8)
  return image, label

def val_images(fn):
  >file_name = bytes.decode(fn.numpy())
  image = tf.io.read_file(os.path.join("val2014/", file_name))
  image = tf.image.decode_image(image, channels=3, dtype=tf.float32)
  image = tf.image.resize(image, (227, 227))
  label = tf.convert_to_tensor(val_annotation.loc[file_name].values, dtype=tf.uint8)
  return image, label

def test_images(fn):
  >file_name = bytes.decode(fn.numpy())
  image = tf.io.read_file(os.path.join("val2014/", file_name))
  image = tf.image.decode_image(image, channels=3, dtype=tf.float32)
  image = tf.image.resize(image, (227, 227))
  label = tf.convert_to_tensor(test_annotation.loc[file_name].values, dtype=tf.uint8)
  return image, label

def shape_function_with_three_dims(func, input, output):
  >holder = tf.py_function(func, [input], output)
  holder[0].set_shape([227, 227, 3])
  holder[1].set_shape([90])
  return holder

Here I create dataset from both situation.
>ds_train = tf.data.Dataset.from_tensor_slices(train_annotation.index.values).map(lambda x: tf.py_function(train_images, [x], [tf.float32, tf.uint8]))
ds_valid = tf.data.Dataset.from_tensor_slices(val_annotation.index.values).map(lambda x: tf.py_function(val_images, [x], [tf.float32, tf.uint8]))
ds_test = tf.data.Dataset.from_tensor_slices(test_annotation.index.values).map(lambda x: tf.py_function(test_images, [x], [tf.float32, tf.uint8]))

And the error that I got: `ValueError: as_list() is not defined on an unknown TensorShape` 


>ds_train1 = tf.data.Dataset.from_tensor_slices(train_annotation.index.values).map(lambda x: shape_function_with_three_dims(train_images, x, [tf.float32, tf.uint8]))
ds_valid1 = tf.data.Dataset.from_tensor_slices(val_annotation.index.values).map(lambda x: shape_function_with_three_dims(val_images, x, [tf.float32, tf.uint8]))
ds_test1 = tf.data.Dataset.from_tensor_slices(test_annotation.index.values).map(lambda x: shape_function_with_three_dims(test_images, x, [tf.float32, tf.uint8]))

And the error when I used the ds_train1 and ds_valid1; `ValueError: Input 0 of layer sequential is incompatible with the layer: : expected min_ndim=4, found ndim=3. Full shape received: (227, 227, 3)`

I also tested to give an input images and label as two different datasets like here:

```
ds_train_images = tf.data.Dataset.from_tensor_slices(train_annotation.index.values).map(lambda x: tf.py_function(train_images, [x], tf.float32))
ds_train_labels = tf.data.Dataset.from_tensor_slices(train_annotation.index.values).map(lambda x: tf.py_function(val_labels, [x], tf.float32))
```

and the inevitable result happened. I got an error: `ValueError: `y` argument is not supported when using dataset as input.`

I just want to create a multi label classifier. If you know how to create a dataset for multi label classifier from the one of the csv file I attach, please share the way with me.
The rows are image id and the columns are the classes for Coco classes. For examle in test_annotation.csv, the image of which id is 558840 includes person, bottle, cup, spoon, hot dog and dining table

[val_annotation.csv](https://github.com/tensorflow/tensorflow/files/6501525/val_annotation.csv)

[train_annotation.csv](https://github.com/tensorflow/tensorflow/files/6501521/train_annotation.csv)

[test_annotation.csv](https://github.com/tensorflow/tensorflow/files/6501520/test_annotation.csv)

**Comment by bhack** [comment]
Can you share a new update colab with mnist so it will be faster to verify?

**Comment by dundarahmet** [comment]
Here the link
[mnist link](https://colab.research.google.com/drive/1GUEQeFm0jK1Vl-GRNh9Kv9Fjj18-5rNU?usp=sharing)
It seems illogical but I tried to keep the same structure as previous link. And I got the same error.

**Comment by bhack** [comment]
Also in this one you need to set the shape as `tf.py_function` don't return tensors with shape and it doesn't have the output_shape extra param e.g. like the generator (here the feature request/proposal https://github.com/tensorflow/tensorflow/issues/36274)

**Comment by bhack** [comment]
If you want an example see how to set the shape with `py_function` at https://www.tensorflow.org/guide/data#applying_arbitrary_python_logic

**Comment by dundarahmet** [comment]
with bhack help, finally my code is working. Thanks dude. But,
I had to add some extra code. These are:

First, I had to create a new method to apple tf.py_function and to give shape to the image and label. Here the method: method takes a function, input-x- and outputs and applies tf.py_function.

>def general_function(func, x, outputs):
 image, label = tf.py_function(train_images, [x], outputs)
  image.set_shape(tf.TensorShape([227, 227, 3]))
  label.set_shape(tf.TensorShape([90]))
  return image, label


After that, when the dataset are created, the batch size must be given using tf.data.Dataset.batch() method. I don't understand why but when I tried to train the model without using .batch() method, but I gave the batch size in model.fit, I got an error -I don't remember which error I got.-

>ds_train = tf.data.Dataset.from_tensor_slices(train_annotation.index.values).map(lambda x: general_function(train_images, x, [tf.float32, tf.uint8])).batch(batch_size)

At last, here my model.fit model's arguments:
>history = model.fit(
    ds_train,
    validation_data=ds_valid,
    epochs=10,
    batch_size=batch_size
)

Note: Here is the train_images method:
>def train_images(fn):
  file_name = bytes.decode(fn.numpy())
  image = tf.io.read_file("train2014/" + file_name)
  image = tf.image.decode_image(image, channels=3, dtype=tf.float32)
  image = tf.image.resize(image, (227, 227))
  label = tf.convert_to_tensor(train_annotation.loc[file_name].values, dtype=tf.uint8)
  return image, label
